[2단계 - 장바구니 협업] 앨버(송상민) 미션 제출합니다.

al-bur opened the pull request:
## step2 장바구니 미션 제출합니다!😀

안녕하세요, 노스님 앨버입니다~

step2 미션 제출합니다. 이번 step에서는 로컬 서버를 백엔드 api로 변경하였습니다. 추가 feature는 개발하지 않았습니다.

<br>

## 실행방법

- 이번에 백엔드와 연결을 하였지만, https 이슈로 인해 배포를 하지 않아 데모 페이지는 존재하지 않습니다. 아래 가이드를 따라 데모 사이트 체크 부탁 드립니다.

1. 레포지토리를 클론한다.

```bash

git clone -b albur-step2 https://github.com/al-bur/react-shopping-cart-prod.git

```

2. 폴더 이동

```bash

cd react-shopping-cart-prod

```

3. 설치 후 실행

```bash

npm i
npm run start

```

<br>


## 구현 사항

- [x] 로컬 서버 -> 백엔드 API 연결

**이번에 루터에 오셨다고 하셨는데...일정이 있어서 뵙지 못했네요ㅠㅠㅠ 아쉽습니다

지금까지 너무 고생하셨고!!! 덕분에 레벨1 레벨2에서 성장할 수있었습니다 감사합니다!!

al-bur committed:
refactor: 사용하지 않는 useState 삭제

관련 피드백: https://github.com/woowacourse/react-shopping-cart-prod/pull/21#discussion_r889195106

al-bur committed:
refactor: prefix 컨벤션에 따라 수정

handle prefix -> on prefix로 수정

관련 피드백: https://github.com/woowacourse/react-shopping-cart-prod/pull/21#discussion_r889213574

al-bur committed:
chore: 자동정렬 설정 추가

al-bur committed:
refactor: 추가한 prettier 전체 파일에 적용

- import 자동 정렬
- , 설정

al-bur committed:
feat: 비로그인시 회원정보수정 페이지 접근 불가능하도록 구현

al-bur committed:
fix: msw 작동 오류 해결

al-bur committed:
refactor: mocking 시작 함수 추출

- index.tsx에서 mocking을 시작하는 로직에 대해서 정확히 알필요없을 것 같아 추상화

al-bur committed:
fix: 무한로딩 일어나는 문제 해결

- msw 내부 설정 문제이기 때문에 공식문서를 참고하여 해당 문제 해결

al-bur committed:
feat: 장바구니 기능 관련하여 인가 적용

- 비로그인 상황에서 장바구니 페이지에 접속하려하거나 장바구니 기능 이용시 로그인 페이지로 이동

al-bur committed:
refactor: 로그인 상태유지 관련 예상되는 버그로 인하여 기능 삭제

- localStroage를 통해 로그인 상태를 유지하게 되면, 10분의 유효기간을 가진 토큰이 유효하지 않을 경우에도 로그인 되었다고 판단이 될 것입니다. 현재의 상황에서는 유효한 토큰인지를 검사해줄 수가 없기 때문에 로그인 상태유지 기능을 삭제하였습니다

al-bur committed:
refactor: 상품 재고에 대한 정보 제거 및 관련 기능 삭제

- 메인 페이지내 상품 재고에 대한 정보 제거 (백엔드와의 회의 이후 결정)로 인해 해당 정보 제거 및 관련 기능 삭제
- 카트에 담겨져있는 상품의 양을 나타내는 변수의 네이밍을 stock으로 했던 것을 quantity로 변경

al-bur committed:
refactor: 백엔드 api 명세에 따른 변수명 변경

- image -> imageUrl

al-bur committed:
refactor: thunk 로직 수정

al-bur committed:
chore: redux logger 설정

al-bur committed:
feat: 로컬 서버 -> api 연결

al-bur committed:
fix: 장바구니 전체 선택시 무한으로 장바구니에 담기는 문제 해결

- 현재 장바구니에 담겨있으면 전체 선택시 checkedCartItems에 담기지 않도록 수정

al-bur committed:
feat: 장바구니 상품들 전체 삭제 api 연결

hyoungnam commented on src/components/SignupForm/SignupForm.tsx:
```diff
@@ -26,40 +26,44 @@ function SignupForm() {
   const [passwordCheckCorrect, setPasswordCheckCorrect] = useState(false);
 
-  const handleIdInput = (e: React.ChangeEvent<HTMLInputElement>) => {
+  const onChangeIdInput = (e: React.ChangeEvent<HTMLInputElement>) => {
```
handle에서 on으로 바뀐이유가 궁금하네요 🤔

al-bur replied:
이 부분은 이전 크루가 작성해놓은 코드의 컨벤션이 `handle` prefix가 아닌 `on` prefix이여서 컨벤션에 맞게 수정해주었었습니다.

hyoungnam commented on src/configs/api.ts:
```diff
@@ -1,11 +1,8 @@
 const axios = Axios.create({
-  baseURL: 'http://localhost:8080/api',
+  baseURL:
+    'http://ec2-13-125-246-80.ap-northeast-2.compute.amazonaws.com:8080/api',
```
리뷰때문에 env를 사용하지 않은것 같지만 추후에는 env로 바꿔주면 좋을것 같네요

al-bur replied:
넵 참고하겠습니다!!

al-bur replied:
그런데,  url을 왜 env로 관리해줘야하는 건지 여쭤봐도 될까요?

development와 production 상황에서 API url를 분기하기 위해서라고 생각하면 될까요?

hyoungnam replied:
네 앨버님이 말씀하신것처럼 분기 처리나 배포설정에서 .env로 관리하는것이 좀 더 관리하기 편하겠죠

hyoungnam commented on src/pages/ProductPage/ProductPage.tsx:
```diff
@@ -1,49 +1,61 @@
           <>
             <StyledImageContainer>
-              <img src={productDetail.image} alt={productDetail.name} />
+              <img src={productDetail.imageUrl} alt={productDetail.name} />
```
외부에서 imageUrl을 가지고 오는경우 잘못된 경로로 인해서 종종 있습니다.
src={productDetail.imageUrl ||  ~} 또는 onerror 프로퍼티를 활용해보는것도 좋을듯하네요
한편  picture태그를 한번 살펴보시는것도 추천합니다:)

al-bur replied:
좋네요 한 번 살펴보도록 하겠습니다!!
이미지가 에러가 났을 때 보여줄 것을 설정해주는 것도 좋겠네요

al-bur replied:
기본이미지 보여주도록 수정하였습니다! 4b68896897c84d698a8e1e0c6756a713f9c15ead

hyoungnam commented on src/pages/ProductPage/ProductPage.tsx:
```diff
@@ -1,49 +1,61 @@
+    e.preventDefault();
+
+    // TODO: 반복되는 로직이므로 함수로 추출해야할듯
+    if (!isLogin()) {
+      navigate(PATH.LOGIN);
+
+      return;
+    }
```
갑자기 로그인페이지로 보내는것보다 안내메시지로 한번 물어보는것도 좋을것 같네요

al-bur replied:
그렇네요 한 번 안내메시지를 보내서 의사를 물어보는 방식으로 리팩토링 해보겠습니다!

al-bur replied:
메인페이지와 상품 상세페이지에서 비로그인 상태일때, 장바구니 기능을 이용을 시도하면 confirm 메시지를 띄우도록 해주었습니다. d4d5587ca600decd2304df874939a0c4ef975a7f

al-bur committed:
chore: 기본 백엔드 API 변경

al-bur committed:
fix: 장바구니 상품 전체 삭제시 문제 해결

al-bur committed:
refactor: 불필요한 console.log 제거

hyoungnam commented on src/components/ProductCard/ProductCard.tsx:
```diff
@@ -1,44 +1,59 @@
+      return;
+    }
+
+    //NOTE: 이렇게 비동기처리 해줘도 되는 걸까?
+    cartAPI.addCartItem({ productId: id, quantity: 1 }).then((res) => {
+      res?.status === 201 && dispatch(getCarts());
+    });
```
어떤점때문에 물음표를 하신걸까요?  카트 정보를 백엔드에서 관리하고 있으면 비동기처리를 해야하지 않나요?? 중복되거나 잘못된 요청이면 그에따른 메세지도 보여주면 좋을것 같네요

al-bur replied:
물음표 타겟은 then을 활용해서 이후의 처리를 해줘도 될까?라는 생각이였습니다!!

> 중복되거나 잘못된 요청이면 그에따른 메세지도 보여주면 좋을것 같네요

동의합니다 try catch문으로 감싸서 에러 핸들링을 해주면 좋을 것 같아요

al-bur replied:
then -> try...catch문 + await으로 변경해주었습니다. 138003b2fe90d8fac326e99d706e742dd4619b1d

hyoungnam reviewed (changes requested):
안녕하세요 앨버~! 루터에서 못뵈서 저도 아쉽고 나중에 기회가되서 뵈었으면 좋겠네요:) 백엔드와 연동하기 위해서 많은 부분 작업을 한게 보이네요.  
메인페이지에 들어갔을때 요청에러가 나타내고  무수히 많은 요청이 계속 갔습니다. 이것때문에 서버에 문제가 있는건지 아님 문제가 이전부터 문제가 있는상태인지는 잘 모르겠네요. 에러상태에서 계속 요청가는건 고쳐야할듯합니다. MSW로 돌리긴하였지만 이부분도 회원가입 후 로그인이 정상적으로 진행되지 않아서 페이지에 접근하지는 못했습니다. 한번 확인 후 코멘트 주시거나 DM 주시면 다시 들어가보도록 할께요. 
코드만 보고 몇가지 코멘트 남겼습니다. 확인 한번 해주세요

al-bur committed:
feat: 이미지 불러오기 실패할 시 기본 이미지 보여주도록 설정

관련 피드백: https://github.com/woowacourse/react-shopping-cart-prod/pull/63#discussion_r895135915

al-bur committed:
feat: 메인페이지 및 상품 상세 페이지에서 장바구니 기능 이용시 로그인 페이지 이동 여부 물어보도록 기능 구현

관련 피드백: https://github.com/woowacourse/react-shopping-cart-prod/pull/63#discussion_r895138772

al-bur committed:
refactor: 메인페이지내 장바구니 추가 로직 리팩토링

- then -> try catch + await 사용으로 변경
- 에러 핸들링을 더 깔끔하게 해주기 위해 try catch문 사용

관련 피드백: https://github.com/woowacourse/react-shopping-cart-prod/pull/63#discussion_r895141665

al-bur committed:
refactor: 컬러 상수화

hyoungnam reviewed (approved):
앨버와의 마지막 리뷰네요. 미션수행하느라 고생하셨습니다. 뭔가 좀 더 도움이되는게 더 없을까 생각해보았다가 미션이 상품과 관련되어 있기도 하여 상품 SEO와 관련되서 알게된 링크 하나 남깁니다. [구조화된 데이터 작동 방식 이해](https://developers.google.com/search/docs/advanced/structured-data/intro-structured-data?hl=ko)
방학 잘보내시고 앞으로도 모든 부분에서 화이팅입니다~!
